contracts/nftauction.js

```diff
@@ -0,0 +1,576 @@
+
+    const params = {};
+    // fee required to set up an auction
+    params.creationFee = '1';
+
+    // percent a bid needs to increase to take the lead
+    params.minBidIncrement = 50; // 5%
+
+    // time remaining in the auction settling when cancel action is locked
+    params.cancelLockTime = 300000; // 5 mins
+
+    // time after the last lead bid it takes to settle the auction
+    params.expiryTime = 86400000; // 24 hours
+
+    // max time an auction can run
+    params.maxExpiryTime = 2592000000; // 30 days
+
+    // max auctions to settle per block
+    params.auctionsPerBlock = 2;
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.assert(api.sender === api.owner, 'not authorized')) {
+    const {
+      creationFee,
+      minBidIncrement,
+      cancelLockTime,
+      expiryTime,
+      maxExpiryTime,
+      auctionsPerBlock,
+    } = payload;
+
+    const params = await api.db.findOne('params', {});
+
+    if (creationFee) {
+      if (!api.assert(typeof creationFee === 'string' && !api.BigNumber(creationFee).isNaN() && api.BigNumber(creationFee).gte(0), 'invalid creationFee')) return;
+      params.creationFee = creationFee;
+    }
+    if (minBidIncrement) {
+      if (!api.assert(Number.isInteger(minBidIncrement) && minBidIncrement > 0, 'invalid minBidIncrement')) return;
+      params.minBidIncrement = minBidIncrement;
+    }
+    if (cancelLockTime) {
+      if (!api.assert(Number.isInteger(cancelLockTime) && cancelLockTime > 0, 'invalid cancelLockTime')) return;
+      params.cancelLockTime = cancelLockTime;
+    }
+    if (expiryTime) {
+      if (!api.assert(Number.isInteger(expiryTime) && expiryTime > 0, 'invalid expiryTime')) return;
+      params.expiryTime = expiryTime;
+    }
+    if (maxExpiryTime) {
+      if (!api.assert(Number.isInteger(maxExpiryTime) && maxExpiryTime > 0, 'invalid maxExpiryTime')) return;
+      params.maxExpiryTime = maxExpiryTime;
+    }
+    if (auctionsPerBlock) {
+      if (!api.assert(Number.isInteger(auctionsPerBlock) && auctionsPerBlock > 0, 'invalid auctionsPerBlock')) return;
+      params.auctionsPerBlock = auctionsPerBlock;
+    }
+
+    await api.db.update('params', params);
+  }
+};
+
+const countDecimals = value => api.BigNumber(value).dp();
+
+const getTimestamp = (value) => {
+  try {
+    const date = new Date(`${value}.00Z`);
+    return date.getTime();
+  } catch (e) {
+    return false;
+  }
+};
+
+const transferIsSuccessful = (result, action, from, to, symbol, quantity) => {
+  if (result.errors === undefined
+    && result.events && result.events.find(el => el.contract === 'tokens'
+    && el.event === action
+    && el.data.from === from
+    && el.data.to === to
+    && api.BigNumber(el.data.quantity).eq(quantity)
+    && el.data.symbol === symbol) !== undefined) {
+    return true;
+  }
+
+  return false;
+};
+
+const settleAuction = async (auction, id = null, settle = true) => {
```

Comment:
settle and cancel are different enough that I think this serves better as splitting into their respective functions and extracting common pieces as helper methods. For example, it looks like you return nfts to the seller twice on cancel.